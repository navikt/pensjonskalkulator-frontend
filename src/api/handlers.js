import { rest } from 'msw'

import pensjonsberegningData from './__mocks__/pensjonsberegning.json' assert { type: 'json' }

// const target = 'http://localhost:8088'
// const apiPath = '/pensjon/kalkulator/api'

export const handlers = [
  rest.get(
    `http://localhost:8088/pensjon/kalkulator/api/pensjonsberegning`,
    (req, res, ctx) => {
      console.log('>>>>> HANDLERS')
      // successful response
      return res(
        ctx.status(200),
        ctx.json(pensjonsberegningData),
        ctx.delay(30)
      )
    }
  ),
]
