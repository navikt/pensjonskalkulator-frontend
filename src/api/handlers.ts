import { rest } from 'msw'

import pensjonsberegningData from './__mocks__/pensjonsberegning.json' assert { type: 'json' }

const target = 'http://localhost:8088'
const apiPath = '/pensjon/kalkulator/api'

export const handlers = [
  rest.get(`${target}${apiPath}/pensjonsberegning`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(pensjonsberegningData), ctx.delay(30))
  }),
]
