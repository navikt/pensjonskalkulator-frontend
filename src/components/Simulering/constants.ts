export const MAX_UTTAKSALDER = 77
export const COLUMN_WIDTH = 25
export const TOOLTIP_YPOS = 35

export const SERIES_COLORS = {
  SERIE_COLOR_INNTEKT: 'var(--a-gray-500)',
  SERIE_COLOR_AFP: 'var(--a-purple-400)',
  SERIE_COLOR_TP: 'var(--a-green-400)',
  SERIE_COLOR_ALDERSPENSJON: 'var(--a-deepblue-500)',
  SERIE_COLOR_FADED_INNTEKT: 'var(--a-gray-300)',
  SERIE_COLOR_FADED_AFP: 'var(--a-purple-200)',
  SERIE_COLOR_FADED_TP: 'var(--a-green-200)',
  SERIE_COLOR_FADED_ALDERSPENSJON: 'var(--a-deepblue-200)',
}

export const SERIES_DEFAULT = {
  SERIE_INNTEKT: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Pensjonsgivende inntekt',
    color: SERIES_COLORS.SERIE_COLOR_INNTEKT,
  },
  SERIE_TP: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Pensjonsavtaler (arbeidsgivere m.m.)',
    color: SERIES_COLORS.SERIE_COLOR_TP,
  },
  SERIE_AFP: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'AFP (Avtalefestet pensjon)',
    color: SERIES_COLORS.SERIE_COLOR_AFP,
  },
  SERIE_ALDERSPENSJON: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Alderspensjon (NAV)',
    color: SERIES_COLORS.SERIE_COLOR_ALDERSPENSJON,
  },
}

export const highchartsScrollingSelector = '.highcharts-scrolling'
