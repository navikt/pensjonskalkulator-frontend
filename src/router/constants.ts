export const BASE_PATH = '/pensjon/kalkulator'

export const externalUrls = {
  dinPensjon: 'https://nav.no/pensjon',
  dinPensjonBeholdning: 'https://www.nav.no/pensjon/opptjening/nb/',
  detaljertKalkulator: 'https://www.nav.no/pselv/simulering.jsf',
  alderspensjonsregler: 'https://www.nav.no/alderspensjon#beregning',
  afp: 'https://www.afp.no',
  afpPrivat: 'https://www.nav.no/afp-i-privat-sektor',
  garantipensjon: 'https://www.nav.no/minstepensjon',
  norskPensjon: 'https://norskpensjon.no/',
  uinnloggetKalkulator:
    'https://www.nav.no/pselv/simulering/forenkletsimulering.jsf',
  personvernerklaering:
    'https://www.nav.no/personvernerklaering#dine-rettigheter',
  personvernerklaeringKontaktOss:
    'https://www.nav.no/personvernerklaering#kontakt-nav',
  kontaktOss: 'https://www.nav.no/kontaktoss',
  planleggePensjon:
    'https://www.nav.no/planlegger-pensjon#noe-du-ikke-finner-svaret-p-her',
}

export const paths = {
  root: '/',
  login: '/login',
  henvisning: '/henvisning',
  start: '/start',
  samtykke: '/samtykke',
  utenlandsopphold: '/utenlandsopphold',
  offentligTp: '/offentlig-tp',
  afp: '/afp',
  samtykkeOffentligAFP: '/samtykke-offentlig-afp',
  ufoeretrygdAFP: '/ufoeretrygd-afp',
  sivilstand: '/sivilstand',
  uventetFeil: '/uventet-feil',
  beregningEnkel: '/beregning',
  beregningAvansert: '/beregning-detaljert',
  forbehold: '/forbehold',
  personopplysninger: '/personopplysninger',
} as const

export const henvisningUrlParams = {
  foedselsdato: '1963',
  apotekerne: 'apotekerne',
  utland: 'utland',
} as const
