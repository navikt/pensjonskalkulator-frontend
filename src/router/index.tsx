export { routes } from './routes'
export { externalUrls, paths, BASE_PATH } from './constants'
