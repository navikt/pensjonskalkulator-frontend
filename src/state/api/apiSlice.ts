import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react'
import { isPensjonsberegning } from './typeguards'

export const apiSlice = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({
    baseUrl: `${import.meta.env.VITE_MSW_BASEURL ?? ''}/pensjon/kalkulator/api`,
  }),
  endpoints: (builder) => ({
    getPensjonsberegning: builder.query<Pensjonsberegning[], void>({
      query: () => '/pensjonsberegning',
      transformResponse: (response: Pensjonsberegning[]) => {
        if (!isPensjonsberegning(response)) {
          throw new Error(`Mottok ugyldig pensjonsberegning: ${response}`)
        }
        return response
      },
    }),
  }),
})

export const { useGetPensjonsberegningQuery } = apiSlice
