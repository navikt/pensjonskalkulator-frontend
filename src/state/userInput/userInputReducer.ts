import { createSlice, PayloadAction } from '@reduxjs/toolkit'

import { formatInntekt } from '../../utils/inntekt'

export interface Simulation {
  formatertUttaksalderReadOnly: string | null // (!) Obs READONLY - string i format "YY alder.aar string.og M alder.maaneder" - oppdateres automatisk basert på uttaksalder - se uttaksalderListener
  uttaksalder: Alder | null // valgt uttaksalder for 100% alderspensjon (alder perioden gjelder FRA) - aar heltall, maaneder heltall mellom 0-11
  aarligInntektFoerUttakBeloep: string | null // inntekt før uttak av pensjon - formatert string i nok - overskriver beløp fra Skatteetaten
  aarligInntektVsaHelPensjon?: AarligInntektVsaPensjon
  gradertUttaksperiode: GradertUttak | null
}

export interface UserInputState {
  utenlandsopphold: boolean | null
  samtykke: boolean | null
  afp: AfpRadio | null
  samboer: boolean | null
  currentSimulation: Simulation
}

export const userInputInitialState: UserInputState = {
  utenlandsopphold: null,
  samtykke: null,
  afp: null,
  samboer: null,
  currentSimulation: {
    formatertUttaksalderReadOnly: null,
    uttaksalder: null,
    aarligInntektFoerUttakBeloep: null,
    gradertUttaksperiode: null,
  },
}

export const userInputSlice = createSlice({
  name: 'userInputSlice',
  initialState: userInputInitialState,
  reducers: {
    setUtenlandsopphold: (state, action: PayloadAction<boolean>) => {
      state.utenlandsopphold = action.payload
    },
    setSamtykke: (state, action: PayloadAction<boolean>) => {
      state.samtykke = action.payload
    },
    setAfp: (state, action: PayloadAction<AfpRadio>) => {
      state.afp = action.payload
    },
    setSamboer: (state, action: PayloadAction<boolean>) => {
      state.samboer = action.payload
    },
    setCurrentSimulationUttaksalder: (
      state,
      action: PayloadAction<{
        aar: number
        maaneder: number
      } | null>
    ) => {
      state.currentSimulation = {
        ...state.currentSimulation,
        uttaksalder: action.payload ? { ...action.payload } : null,
      }
    },
    setCurrentSimulationAarligInntektFoerUttakBeloep: (
      state,
      action: PayloadAction<string | null>
    ) => {
      state.currentSimulation.aarligInntektFoerUttakBeloep =
        action.payload !== null ? formatInntekt(action.payload) : null
    },
    setCurrentSimulationAarligInntektVsaHelPensjon: (
      state,
      action: PayloadAction<AarligInntektVsaPensjon | undefined>
    ) => {
      state.currentSimulation.aarligInntektVsaHelPensjon =
        action.payload && action.payload.beloep
          ? {
              ...action.payload,
              beloep: formatInntekt(action.payload?.beloep),
            }
          : undefined
    },
    setCurrentSimulationGradertUttaksperiode: (
      state,
      action: PayloadAction<GradertUttak | null>
    ) => {
      state.currentSimulation.gradertUttaksperiode = action.payload
        ? {
            ...action.payload,
            aarligInntektVsaPensjonBeloep: action.payload
              ? formatInntekt(action.payload.aarligInntektVsaPensjonBeloep)
              : undefined,
          }
        : null
    },
    syncCurrentSimulationFormatertUttaksalderReadOnly: (
      state,
      action: PayloadAction<string | null>
    ) => {
      state.currentSimulation = {
        ...state.currentSimulation,
        formatertUttaksalderReadOnly: action.payload,
      }
    },
    flush: (state) => {
      state.utenlandsopphold = null
      state.samtykke = null
      state.afp = null
      state.samboer = null
      state.currentSimulation = { ...userInputInitialState.currentSimulation }
    },
    flushCurrentSimulation: (state) => {
      state.currentSimulation = { ...userInputInitialState.currentSimulation }
    },
  },
})

export default userInputSlice.reducer

export const userInputActions = userInputSlice.actions

export type UserInputSlice = {
  [userInputSlice.name]: ReturnType<(typeof userInputSlice)['reducer']>
}
