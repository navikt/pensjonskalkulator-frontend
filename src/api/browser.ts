import { setupWorker } from 'msw'
import { getHandlers } from './handlers'
import { apiPath } from './server'

const handlers = getHandlers(apiPath)
export const worker = setupWorker(...handlers)
