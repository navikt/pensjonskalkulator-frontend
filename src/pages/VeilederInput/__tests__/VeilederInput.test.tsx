import { describe, it } from 'vitest'

import { VeilederInput } from '../'
import { BASE_PATH, paths } from '@/router/constants'
import * as apiSliceUtils from '@/state/api/apiSlice'
import { userInputInitialState } from '@/state/userInput/userInputReducer'
import * as userInputReducerUtils from '@/state/userInput/userInputReducer'
import {
  render,
  screen,
  waitFor,
  userEvent,
  swallowErrorsAsync,
} from '@/test-utils'

const previousWindow = window

describe('VeilederInput', () => {
  afterEach(() => {
    vi.clearAllMocks()
    vi.resetAllMocks()
    global.window = previousWindow
  })

  describe('Gitt at veilederen besøker en side som er ekskludert', () => {
    it('Når veilederen klikker på tittelen, vises det ansattid med vanlig side under', async () => {
      const user = userEvent.setup()
      global.window = Object.create(window)
      Object.defineProperty(window, 'location', {
        value: {
          href: 'before',
          pathname: `/veileder${paths.forbehold}`,
        },
        writable: true,
      })
      render(<VeilederInput />, {
        hasRouter: false,
      })
      expect(window.location.href).toBe('before')

      await user.click(screen.getByText('Pensjonskalkulator', { exact: true }))
      expect(window.location.href).toBe(`${BASE_PATH}/veileder`)
      expect(screen.getByTestId('veileder-ekskludert-side')).toBeVisible()
      expect(screen.getByText('ABC123')).toBeVisible()
    })
  })

  describe('Gitt at borger ikke er valgt', () => {
    it('viser sjema med input og submit knapp', () => {
      render(<VeilederInput />, { hasRouter: false })
      expect(screen.getByTestId('borger-fnr-input')).toBeInTheDocument()
    })

    it('Når borger fnr tastes inn og submit knappen trykkes på, lagres fnr og encrypted fnr i store', async () => {
      const setVeilederBorgerFnrMock = vi.spyOn(
        userInputReducerUtils.userInputActions,
        'setVeilederBorgerFnr'
      )
      const invalidateMock = vi.spyOn(
        apiSliceUtils.apiSlice.util.invalidateTags,
        'match'
      )
      const user = userEvent.setup()
      render(<VeilederInput />, { hasRouter: false })
      expect(screen.getByText('Veiledertilgang')).toBeVisible()
      const submitButton = screen.getByTestId('veileder-submit')
      expect(submitButton).toHaveTextContent('Logg inn')

      const input = screen.getByTestId('borger-fnr-input')
      await user.clear(input)
      await user.type(input, '1')
      await user.type(input, '0')
      await user.type(input, '0')
      await user.type(input, '3')
      await user.type(input, '6')
      await user.type(input, '5')
      await user.type(input, '9')
      await user.type(input, '9')
      await user.type(input, '9')
      await user.type(input, '9')
      await user.type(input, '9')

      await user.click(submitButton)

      expect(setVeilederBorgerFnrMock).toHaveBeenCalledWith({
        encryptedFnr: 'this-is-just-jibbrish-encrypted-fnr',
        fnr: '10036599999',
      })
      expect(invalidateMock).toHaveBeenCalled()
    })

    it('Når en feil oppstår ved fnr encription, vises det riktig feilmelding', async () => {
      const fetchMock = vi
        .spyOn(global, 'fetch')
        .mockResolvedValueOnce({
          // headers: Headers,
          ok: true,
          status: 200,
          statusText: 'fullfilled',
          type: 'default',
          url: 'http://localhost:8088/pensjon/kalkulator/oauth2/session',
          redirected: false,
          // clone: () => Resp
          //  error (): Response
          // json(data: any, init?: ResponseInit): Response
          // redirect (url: string | URL, status: ResponseRedirectStatus): Response } as unknown as Response)
        } as Response)
        .mockResolvedValue({
          // headers: Headers,
          ok: false,
          status: 503,
          statusText: 'error',
          type: 'default',
          url: 'http://localhost:8088/pensjon/kalkulator/v1/encrypt',
          redirected: false,
          // clone: () => Resp
          //  error (): Response
          // json(data: any, init?: ResponseInit): Response
          // redirect (url: string | URL, status: ResponseRedirectStatus): Response } as unknown as Response)
        } as Response)

      const user = userEvent.setup()

      await swallowErrorsAsync(async () => {
        render(<VeilederInput />, { hasRouter: false })
        expect(screen.getByText('Veiledertilgang')).toBeVisible()
        const submitButton = screen.getByTestId('veileder-submit')
        expect(submitButton).toHaveTextContent('Logg inn')

        const input = screen.getByTestId('borger-fnr-input')
        await user.clear(input)
        await user.type(input, '1')
        await user.type(input, '0')
        await user.type(input, '0')
        await user.type(input, '3')
        await user.type(input, '6')
        await user.type(input, '5')
        await user.type(input, '9')
        await user.type(input, '9')
        await user.type(input, '9')
        await user.type(input, '9')
        await user.type(input, '9')

        await user.click(submitButton)
        expect(fetchMock).toHaveBeenCalledTimes(3)
        expect(fetchMock).toHaveBeenNthCalledWith(
          1,
          'http://localhost:8088/pensjon/kalkulator/oauth2/session'
        )
        expect(fetchMock).toHaveBeenNthCalledWith(
          3,
          'http://localhost:8088/pensjon/kalkulator/api/v1/encrypt',
          {
            body: '10036599999',
            method: 'POST',
          }
        )
      })
    })
  })

  describe('Gitt at borger er valgt', () => {
    it('viser kalkulatoren', async () => {
      render(<VeilederInput />, {
        hasRouter: false,
        preloadedState: {
          userInput: {
            ...userInputInitialState,
            veilederBorgerFnr: '12345678901',
          },
        },
      })
      await waitFor(async () => {
        expect(screen.queryByTestId('borger-fnr-input')).not.toBeInTheDocument()
      })
    })

    it('Når veilederen klikker på tittelen', async () => {
      const user = userEvent.setup()
      global.window = Object.create(window)
      Object.defineProperty(window, 'location', {
        value: {
          href: 'before',
          pathname: 'before',
        },
        writable: true,
      })
      render(<VeilederInput />, {
        hasRouter: false,
        preloadedState: {
          userInput: {
            ...userInputInitialState,
            veilederBorgerFnr: '12345678901',
          },
        },
      })
      expect(window.location.href).toBe('before')

      await user.click(screen.getByText('Pensjonskalkulator', { exact: true }))
      expect(window.location.href).toBe(`${BASE_PATH}/veileder`)
    })

    it('viser inaktivet alert', async () => {
      vi.spyOn(URLSearchParams.prototype, 'has').mockImplementation(() => true)
      render(<VeilederInput />, {
        hasRouter: false,
      })
      await waitFor(async () => {
        expect(screen.getByTestId('inaktiv-alert')).toBeInTheDocument()
      })
    })
  })
})
