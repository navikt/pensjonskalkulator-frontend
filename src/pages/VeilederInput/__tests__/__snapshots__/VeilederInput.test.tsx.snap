// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VeilederInput > Gitt at brukeren besøker en side som er ekskludert > Når brukeren klikker på tittelen 1`] = `
<DocumentFragment>
  <div>
    <header
      class="navds-internalheader"
      data-theme="dark"
    >
      <a
        class="navds-internalheader__title navds-body-short"
      >
        <span>
          Pensjonskalkulator
        </span>
      </a>
      <span
        class="navds-stack__spacer"
      />
      <div
        class="navds-internalheader__user"
      >
        <span>
          <div
            class="navds-body-short navds-body-short--small"
          >
            ABC123
          </div>
        </span>
      </div>
    </header>
  </div>
</DocumentFragment>
`;

exports[`VeilederInput > Gitt at veilederen besøker en side som er ekskludert > Når veilederen klikker på tittelen, vises det ansattid med vanlig side under 1`] = `
<DocumentFragment>
  <div>
    <header
      class="navds-internalheader"
      data-theme="dark"
    >
      <a
        class="navds-internalheader__title navds-body-short"
      >
        <span>
          Pensjonskalkulator
        </span>
      </a>
      <span
        class="navds-stack__spacer"
      />
      <div
        class="navds-internalheader__user"
      >
        <span>
          <div
            class="navds-body-short navds-body-short--small"
          >
            ABC123
          </div>
        </span>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
